c2rust-analyze/src/rewrite/expr/unlower.rs

```diff
@@ -0,0 +1,379 @@
+        _rv: &'a mir::Rvalue<'tcx>,
+        _extra_locs: &[Location],
+    ) {
+        let _g = panic_detail::set_current_span(ex.span);
+        // TODO: handle adjustments, overloaded operators, etc
+    }
+
+    fn visit_expr_operand(
+        &mut self,
+        ex: &'tcx hir::Expr<'tcx>,
+        _rv: &'a mir::Operand<'tcx>,
+        _extra_locs: &[Location],
+    ) {
+        let _g = panic_detail::set_current_span(ex.span);
+        // TODO: handle adjustments, overloaded operators, etc
+    }
+}
+
+impl<'a, 'tcx> intravisit::Visitor<'tcx> for UnlowerVisitor<'a, 'tcx> {
+    type NestedFilter = nested_filter::OnlyBodies;
+
+    fn nested_visit_map(&mut self) -> Self::Map {
+        self.tcx.hir()
+    }
+
+    fn visit_expr(&mut self, ex: &'tcx hir::Expr<'tcx>) {
+        self.visit_expr_inner(ex);
+        intravisit::walk_expr(self, ex);
+    }
+}
+
+pub fn unlower<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    mir: &Body<'tcx>,
+    hir_body_id: hir::BodyId,
+) -> BTreeMap<(Location, Vec<SubLoc>), MirOrigin> {
+    // If this MIR body came from a `#[derive]`, ignore it.
+    let def_id = mir.source.def_id();
+    if let Some(parent_def_id) = tcx.opt_parent(def_id) {
+        if tcx.has_attr(parent_def_id, sym::automatically_derived) {
+            return BTreeMap::new();
+        }
+    }
+
+    // Build `span_index`, which maps `Span`s to MIR `Locations`.
+    let span_index = build_span_index(mir);
+
+    // Run the visitor.
+    let hir = tcx.hir().body(hir_body_id);
+
+    let mut v = UnlowerVisitor {
+        tcx,
+        mir,
+        span_index,
+        mir_map: BTreeMap::new(),
+    };
+    intravisit::Visitor::visit_body(&mut v, hir);
+
+    // Print results.
+    eprintln!("unlowering for {:?}:", mir.source);
+    for (bb_id, bb) in mir.basic_blocks().iter_enumerated() {
+        eprintln!("  block {:?}:", bb_id);
+        for (i, stmt) in bb.statements.iter().enumerate() {
+            let loc = Location {
+                block: bb_id,
+                statement_index: i,
+            };
+
+            eprintln!("    {:?}: {:?}", loc, stmt);
+            for (k, v) in v.mir_map.range(&(loc, vec![])..) {
+                if k.0 != loc {
+                    break;
+                }
+                let ex = tcx.hir().expect_expr(v.hir_id);
+                eprintln!("      {:?}: {:?}, {:?}", k.1, v.desc, ex.span);
+            }
+        }
+
+        {
+            let term = bb.terminator();
+            let loc = Location {
+                block: bb_id,
+                statement_index: bb.statements.len(),
+            };
+
+            eprintln!("    {:?}: {:?}", loc, term);
+            for (k, v) in v.mir_map.range(&(loc, vec![])..) {
+                if k.0 != loc {
+                    break;
+                }
+                let ex = tcx.hir().expect_expr(v.hir_id);
+                eprintln!("      {:?}: {:?}, {:?}", k.1, v.desc, ex.span);
+            }
+        }
+    }
+
+    v.mir_map
```

Comment:
This seems like a perfect case for a `Display` `impl`.

```suggestion
    let unlowering = UnloweringMap {
        tcx,
        mir,
        map: visitor.mir_map,
    };

    // Print results.
    eprintln!("{}", unlowering);

    unlowering.map
```

along with

```rust
struct UnloweringMap<'tcx, 'a> {
    tcx: TyCtxt<'tcx>,
    mir: &'a Body<'tcx>,
    map: BTreeMap<(Location, Vec<SubLoc>), MirOrigin>,
}

impl Display for UnloweringMap<'_, '_> {
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        writeln!(f, "unlowering for {:?}:", self.mir.source)?;
        for (bb_id, bb) in self.mir.basic_blocks().iter_enumerated() {
            writeln!(f, "  block {:?}:", bb_id)?;
            for (i, stmt) in bb.statements.iter().enumerate() {
                let loc = Location {
                    block: bb_id,
                    statement_index: i,
                };

                writeln!(f, "    {:?}: {:?}", loc, stmt)?;
                for (k, v) in self.map.range(&(loc, vec![])..) {
                    if k.0 != loc {
                        break;
                    }
                    let ex = self.tcx.hir().expect_expr(v.hir_id);
                    writeln!(f, "      {:?}: {:?}, {:?}", k.1, v.desc, ex.span)?;
                }
            }

            {
                let term = bb.terminator();
                let loc = Location {
                    block: bb_id,
                    statement_index: bb.statements.len(),
                };

                writeln!(f, "    {:?}: {:?}", loc, term)?;
                for (k, v) in self.map.range(&(loc, vec![])..) {
                    if k.0 != loc {
                        break;
                    }
                    let ex = self.tcx.hir().expect_expr(v.hir_id);
                    writeln!(f, "      {:?}: {:?}, {:?}", k.1, v.desc, ex.span)?;
                }
            }
        }
        Ok(())
    }
}
```

Reply:
I also think using `{var:?}`-style format strings would make it easier to match the code here with the docs using the output of this, which would then make it easier to understand which things are meant to be which in the docs.

Reply:
>it'd be more easily useable from any other potential contexts.

I don't think we're likely to want this exact debug output in any other contexts.

>using `{var:?}`-style format strings would

Done in ddf3da625fa2f883a35b0766f48767ef32d21c10

Reply:
> I don't think we're likely to want this exact debug output in any other contexts.

It could be helpful in debugging.  If you feel strongly that it shouldn't be a `Display` `impl`, though, I'd still prefer for it to at least be separated into its own function, both to divide the cognitive complexity of `fn unlower` and to intra-doc link to it from where it's output is shown in the docs.

Reply:
I've factored it out, though IMO this makes little difference for readability (anyone uninterested in the printing of results could already skip over the section clearly labeled "print results")